### ISSUE TITLE
Syntax error in Switch.pm

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7245]
Migrated from [rt.perl.org#28966](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=28966) (status was 'resolved')

Searchable as RT28966$

**Comment by p5pRT** [comment]
### From lool@via.ecp.fr

#### Created by lool@via.ecp.fr



&nbsp;   Hi\,

 User of the Switch package\, I noticed a strange behavior with very
 simple regexps\.

 When I try to run this simple program&#8203;:

&nbsp;    \#\!/usr/bin/perl \-w
&nbsp;    use strict;
&nbsp;    use Switch;
&nbsp;    switch \('foo'\) \{
&nbsp;       case /^bar\\s\+\\S\+/ \{
&nbsp;           \# \+
&nbsp;       \}
&nbsp;    \}
 \(please note the comment\)
 \.\.\. I get&#8203;:
&nbsp;    syntax error at \./buggy\.pl line 7\, near "\}continue"
&nbsp;    syntax error at \./buggy\.pl line 8\, near "\}"
&nbsp;    Execution of \./buggy\.pl aborted due to compilation errors\.

 Removing the comment stops the barking\.

 This does not happen in comments only\, but in any piece of text with
 some "\+"ses here and there\.

 Sorry\, I'd like to provide a patch or to point the culprit\, but this
 is beyond my skills\.

&nbsp;   Regards\,

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.3:

Configured by Debian Project at Sat Mar 27 17:07:14 EST 2004.

Summary of my perl5 (revision 5.0 version 8 subversion 3) configuration:
  Platform:
    osname=linux, osvers=2.4.25-ti1211, archname=i386-linux-thread-multi
    uname='linux kosh 2.4.25-ti1211 #1 thu feb 19 18:20:12 est 2004 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.3 -Dsitearch=/usr/local/lib/perl/5.8.3 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.3 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.3.3 (Debian 20040314)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so.5.8.3
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.3:
    /etc/perl
    /usr/local/lib/perl/5.8.3
    /usr/local/share/perl/5.8.3
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.3:
    HOME=/home/lool
    LANG=fr_FR@euro
    LANGUAGE (unset)
    LC_ADDRESS=fr_FR@euro
    LC_ALL=
    LC_COLLATE=fr_FR@euro
    LC_CTYPE=fr_FR@euro
    LC_IDENTIFICATION=fr_FR@euro
    LC_MEASUREMENT=fr_FR@euro
    LC_MESSAGES=fr_FR@euro
    LC_MONETARY=fr_FR@euro
    LC_NAME=fr_FR@euro
    LC_NUMERIC=fr_FR@euro
    LC_PAPER=fr_FR@euro
    LC_TELEPHONE=fr_FR@euro
    LC_TIME=fr_FR@euro
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/lool/bin/:/home/lool/bin/kernel/:/home/lool/bin/zsh/:/usr/local/sbin:/sbin:/usr/sbin:/usr/local/bin:/bin:/usr/bin:/usr/bin/X11:/usr/games:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Apr 19\, 2004 at 05&#8203;:20&#8203;:34PM \-0000\, Loic Minier wrote&#8203;:
>  User of the Switch package\, I noticed a strange behavior with very
>  simple regexps\.
> 
>  When I try to run this simple program&#8203;:
> 
>     \#\!/usr/bin/perl \-w
>     use strict;
>     use Switch;
>     switch \('foo'\) \{
>        case /^bar\\s\+\\S\+/ \{
>            \# \+
>        \}
>     \}
>  \(please note the comment\)
>  \.\.\. I get&#8203;:
>     syntax error at \./buggy\.pl line 7\, near "\}continue"
>     syntax error at \./buggy\.pl line 8\, near "\}"
>     Execution of \./buggy\.pl aborted due to compilation errors\.
> 
>  Removing the comment stops the barking\.

Thanks for the report\, and sorry for the delay in replying\. The short
answer is that its fixed by the patch below to Text&#8203;::Balanced that's been
applied to the development version of Perl\. Whether and how this fix makes
it into the CPAN version is up to the module's maintainer\, Damian\.
Damian?

The long answer&#8203;:
Switch uses Text&#8203;::Balanced&#8203;::\_match\_codeblock to extract out the block of
text corresponding to

&nbsp;    \{
&nbsp;       case /^bar\\s\+\\S\+/ \{
&nbsp;           \# \+
&nbsp;       \}
&nbsp;    \}

but it runs into problems with the bareword 'case' followed by the bare
pattern match //\. In general\, does

&nbsp;    foo /2 \.\.\.

indicate

call foo\(\) then divide by 2\, then \.\.\.
or match $\_ against /2\.\.\./ and pass the result to foo\(\)?

How Perl interprets this depends upon such things as whether foo is a
builtin\, or has been previously declared\, or whether its a constant
sub\, or whether its an indirect object like a filehandle in a print etc
etc\.

In general\, \_match\_codeblock\(\) can't possibly disambiguate this; at the
moment it assumes the thing following is an operator except for
a few special\-cased barewords like split and grep\. I've added 'case'
to this list of exceptions\. This menas it now sees the /\.\.\./ thing
following the 'case' as a regex to be skipped over\.

Dave\.



\-\- 
Wesley Crusher gets beaten up by his classmates for being a smarmy git\,
and consequently has a go at making some friends of his own age for a
change\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#18


Change 22821 by davem@&#8203;davem\-percy on 2004/05/16 15&#8203;:31&#8203;:42

&nbsp;    make Text&#8203;::Balanced skip "case /\.\.\.\./" correctly for Switch\.pm

Affected files \.\.\.

\.\.\. //depot/perl/lib/Text/Balanced\.pm\#9 edit
\.\.\. //depot/perl/lib/Text/Balanced/t/extcbk\.t\#4 edit

Differences \.\.\.

==== //depot/perl/lib/Text/Balanced\.pm\#9 \(text\) ====

@&#8203;@&#8203; \-10\,7 \+10\,7 @&#8203;@&#8203;
 use SelfLoader;
 use vars qw \{ $VERSION @&#8203;ISA %EXPORT\_TAGS \};
 
\-$VERSION = '1\.95';
\+$VERSION = '1\.95\_01';
 @&#8203;ISA        = qw \( Exporter \);
&nbsp;              
 %EXPORT\_TAGS    = \( ALL => \[ qw\(
@&#8203;@&#8203; \-583\,12 \+583\,15 @&#8203;@&#8203;
 
 
&nbsp;         \# NEED TO COVER MANY MORE CASES HERE\!\!\!
\+        \# NB 'case' is included here\, because in Switch\.pm\,
\+        \# it's followed by a term\, not an op
\+
&nbsp;         if \($$textref =~ m\#\\G\\s\*\(?\!$ldel\_inner\)
&nbsp;                     \( \[\-\+\*x/%^&|\.\]=?
&nbsp;                     | \[\!=\]~
&nbsp;                     | =\(?\!>\)
&nbsp;                     | \(\\\*\\\*|&&|\\|\\||\<\<|>>\)=?
\-                    | split|grep|map|return
\+                    | case|split|grep|map|return
&nbsp;                     | \[\(\[\]
&nbsp;                     \)\#gcx\)
&nbsp;         \{

==== //depot/perl/lib/Text/Balanced/t/extcbk\.t\#4 \(text\) ====

@&#8203;@&#8203; \-13\,7 \+13\,7 @&#8203;@&#8203;
 \# Change 1\.\.1 below to 1\.\.last\_test\_to\_print \.
 \# \(It may become useful if the test is moved to \./t subdirectory\.\)
 
\-BEGIN \{ $| = 1; print "1\.\.41\\n"; \}
\+BEGIN \{ $| = 1; print "1\.\.43\\n"; \}
 END \{print "not ok 1\\n" unless $loaded;\}
 use Text&#8203;::Balanced qw \( extract\_codeblock \);
 $loaded = 1;
@&#8203;@&#8203; \-64\,6 \+64\,7 @&#8203;@&#8203;
 
 \# USING&#8203;: extract\_codeblock\($str\);
 \{ $data\[4\] =~ /\['"\]/; \};
\+\{ case /^bar\\s\+\\S\+/ \{\\n\#\+\\n\}\};
 
 \# USING&#8203;: extract\_codeblock\($str\,'\<>'\);
 \< %x = \( try => "this"\) >;

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tux

On Sun 16 May 2004 17&#8203;:57\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> >  \(please note the comment\)
> >  \.\.\. I get&#8203;:
> >     syntax error at \./buggy\.pl line 7\, near "\}continue"
> >     syntax error at \./buggy\.pl line 8\, near "\}"
> >     Execution of \./buggy\.pl aborted due to compilation errors\.
> > 
> >  Removing the comment stops the barking\.
> 
> Thanks for the report\, and sorry for the delay in replying\. The short
> answer is that its fixed by the patch below to Text&#8203;::Balanced that's been
> applied to the development version of Perl\. Whether and how this fix makes
> it into the CPAN version is up to the module's maintainer\, Damian\.
> Damian?

I /think/ someone else took over P&#8203;::R and T&#8203;::B a few month back

\-\- 
H\.Merijn Brand        Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using perl\-5\.6\.1\, 5\.8\.3\, & 5\.9\.x\, and 809 on  HP\-UX 10\.20 & 11\.00\, 11i\,
&nbsp;   AIX 4\.3\, SuSE 9\.0\, and Win2k\.           http&#8203;://www\.cmve\.net/~merijn/
http&#8203;://archives\.develooper\.com/daily\-build@&#8203;perl\.org/   perl\-qa@&#8203;perl\.org
send smoke reports to&#8203;: smokers\-reports@&#8203;perl\.org\, QA&#8203;: http&#8203;://qa\.perl\.org

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, May 16\, 2004 at 09&#8203;:54&#8203;:30PM \+0200\, H\.Merijn Brand wrote&#8203;:
> On Sun 16 May 2004 17&#8203;:57\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> > Thanks for the report\, and sorry for the delay in replying\. The short
> > answer is that its fixed by the patch below to Text&#8203;::Balanced that's been
> > applied to the development version of Perl\. Whether and how this fix makes
> > it into the CPAN version is up to the module's maintainer\, Damian\.
> > Damian?
> 
> I /think/ someone else took over P&#8203;::R and T&#8203;::B a few month back

In which case Porting/Maintainers\.pl and www\.cpan\.org need updating\.\.\.\.

\-\- 
Red sky at night \- gerroff my land\!
Red sky at morning \- gerroff my land\!
&nbsp;    \-\- old farmers' sayings \#14

**Comment by p5pRT** [comment]
### From @rgs

Dave Mitchell wrote&#8203;:
> On Sun\, May 16\, 2004 at 09&#8203;:54&#8203;:30PM \+0200\, H\.Merijn Brand wrote&#8203;:
> > On Sun 16 May 2004 17&#8203;:57\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> > > Thanks for the report\, and sorry for the delay in replying\. The short
> > > answer is that its fixed by the patch below to Text&#8203;::Balanced that's been
> > > applied to the development version of Perl\. Whether and how this fix makes
> > > it into the CPAN version is up to the module's maintainer\, Damian\.
> > > Damian?
> > 
> > I /think/ someone else took over P&#8203;::R and T&#8203;::B a few month back
> 
> In which case Porting/Maintainers\.pl and www\.cpan\.org need updating\.\.\.\.

Done \(change \#22822\); the new maintainer is David Manura\, who I'm cc&#8203;:ing\.

David\, could you look at change 22821 to Text&#8203;::Balanced \(as it's shipped in
bleadperl\) to keep it in sync with the CPAN version? It can be found there&#8203;:

http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse?patch=22821

**Comment by p5pRT** [comment]
### From dm.list@math2.org

Rafael\,

I've incorporated change \#22821 into the development Text&#8203;::Balanced code and 
will monitor bleedperl before releasing anything\.

The current beta version of Text&#8203;::Balanced 1\.96 may be accessed at this project 
page&#8203;:

&nbsp;   http&#8203;://math2\.org/david/text\-balanced/

As shown in the change log\, lots of changes have been made between 1\.95 and 
1\.96\, and most known bugs that can be fixed are fixed\.  This can be released 
shortly to CPAN\.  Testing and feedback is much appreciated\.

best regards\,
\-davidm


Rafael Garcia\-Suarez wrote&#8203;:
> Dave Mitchell wrote&#8203;:
> 
>>On Sun\, May 16\, 2004 at 09&#8203;:54&#8203;:30PM \+0200\, H\.Merijn Brand wrote&#8203;:
>>
>>>On Sun 16 May 2004 17&#8203;:57\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
>>>
>>>>Thanks for the report\, and sorry for the delay in replying\. The short
>>>>answer is that its fixed by the patch below to Text&#8203;::Balanced that's been
>>>>applied to the development version of Perl\. Whether and how this fix makes
>>>>it into the CPAN version is up to the module's maintainer\, Damian\.
>>>>Damian?
>>>
>>>I /think/ someone else took over P&#8203;::R and T&#8203;::B a few month back
>>
>>In which case Porting/Maintainers\.pl and www\.cpan\.org need updating\.\.\.\.
> 
> 
> Done \(change \#22822\); the new maintainer is David Manura\, who I'm cc&#8203;:ing\.
> 
> David\, could you look at change 22821 to Text&#8203;::Balanced \(as it's shipped in
> bleadperl\) to keep it in sync with the CPAN version? It can be found there&#8203;:
> 
> http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse?patch=22821
>

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
